server/src/d2Ast.ts

```diff
@@ -0,0 +1,202 @@
+/**
+ * Used to take the ast JSON that is produced by the D2
+ * command line tool and form it into something the
+ * autocompletion and definition/reference/rename
+ * capabilities can use.
+ */
+
+import {
+  Diagnostic,
+  DiagnosticSeverity,
+  LSPAny,
+  PublishDiagnosticsParams,
+} from "vscode-languageserver";
+
+import { d2ExternalLink, d2Node, d2Range, d2StringAndRange } from "./dataContainers";
+import { Position } from "vscode-languageserver-textdocument";
+import { connection } from "./server";
+
+/**
+ * AstReader class.  Takes the returned json text and
+ * forms usable data for the language server
+ */
+export class AstReader {
+  constructor(astStr: string) {
+    try {
+      this.d2Info = JSON.parse(astStr);
+    } catch (err) {
+      connection.console.error("Possilble D2/VsCode version missmatch.");
+      connection.console.error(astStr);
+    }
+
+    this.range = this.d2Info?.Ast.range;
+
+    this.doNodes(this.d2Info?.Ast.nodes);
+
+    // console.info(JSON.stringify(this.d2Info, null, 2));
+  }
+
+  /**
+   * Storage for class data
+   */
+  private range: d2Range;
+  private d2Info: LSPAny;
+  private nodes: d2Node[] = [];
+  private references: d2StringAndRange[] | undefined;
+
+  /**
+   * Returns errors for the 'problems' window from any
+   * errors produced during AST creation. This leaves a
+   * blind spot for errors that are created during IR,
+   * since that phase isn't used by the language server
+   * to save compute time.
+   */
+  get Errors(): PublishDiagnosticsParams | undefined {
+    const diags: Diagnostic[] = [];
+
+    let hasErrors = false;
```

Comment:
don't need this, `diags.length > 0`

Reply:
The return uses the hasErrors var in order to avoid multiple return paths.